### PR TITLE
Edit field

### PR DESCRIPTION
**Pull request by mahimavuppuluri**
Edit Modal opens on click of edit button in the fields info table. On submit the values are saved and table data is refreshed.
Screenshot 2020-12-11 at 3 31 54 PM

### CODE REVIEW

**Comment by aashnaPrakash** [comment]
- The locality filter is not getting preserved on editing.

- The modal does not appear to be in the center of the screen.

File: src/Pages/FieldInfoPage.js [hunk 1]
```diff
@@ -14,6 +14,8 @@ import React, { useState, useEffect } from 'react';
 import { getAllFields, getFieldsByLocation } from 'dataclients/FieldsClient';
 import { getLocations } from 'dataclients/LocationsClient';
 import TableComponent from 'components/common/TableComponent';
+import SimpleModal from 'components/common/SimpleModal';
```
Comment (gautamguptabasant) [inline src/Pages/FieldInfoPage.js:16]: Can name this `Modal` instead.

File: src/Pages/FieldInfoPage.js [hunk 2]
```diff
@@ -82,7 +84,10 @@ function FieldInfoPage() {
   const [alertMessage, setAlertMessage] = useState('');
   const [alertStatus, setAlertStatus] = useState(false);
   const [alertSeverity, setAlertSeverity] = useState('');
-  const [fieldsData, setFields] = useState([]);
+  const [fieldsData, setFieldData] = useState([]);
```
Comment (gautamguptabasant) [inline src/Pages/FieldInfoPage.js]: setField`s`Data to be consistent

File: src/Pages/FieldInfoPage.js [hunk 3]
```diff
@@ -101,7 +106,7 @@ function FieldInfoPage() {
   const getFieldData = () => {
     getAllFields()
       .then((fields) => {
-        setFields(fields.content);
+        setFieldData(fields.content);
       })
       .catch((e) => {
         console.log('Fetching all fields failed', e);
```
Comment (gautamguptabasant) [inline src/Pages/FieldInfoPage.js:111]: Can you please add showAlert here.

File: src/Pages/FieldInfoPage.js [hunk 4]
```diff
@@ -113,20 +118,26 @@ function FieldInfoPage() {
...
    *
    */
-  const getRowData = () => {
+  const updateRowData = () => {
     let data = fieldsData.map((obj) => {
       return {
         id: obj.id,
         name: obj.identifier,
         location: obj.location.displayStr,
+        locationId: obj.location.code,
```
Comment (gautamguptabasant) [inline src/Pages/FieldInfoPage.js]: `locationId` -> `locationCode` please

File: src/Pages/FieldInfoPage.js [hunk 5]
```diff
@@ -113,20 +118,26 @@ function FieldInfoPage() {
...
         area: obj.area,
+        obj: obj,
```
Comment (gautamguptabasant) [inline src/Pages/FieldInfoPage.js]: Let's name the field something meaningful instead. Suggestion: `field` or `fieldData`

File: src/Pages/FieldInfoPage.js [hunk 6]
```diff
@@ -236,10 +254,23 @@ function FieldInfoPage() {
...
+      <SimpleModal
+        isOpen={isEditModalOpen}
+        closeHandler={handleClose}
+        modalBody={
+          <FieldForm
+            operation='UPDATE'
+            title='Edit Field'
+            selectedRow={selectedRow}
+            isOpen={isEditModalOpen}
+            closeHandler={handleClose}
```
Comment (gautamguptabasant) [inline src/Pages/FieldInfoPage.js:266]: Ideally, we should not need to pass these to both the SimpleModal and the inner component.
Let's see if we can improve on this.

File: src/components/common/FieldForm.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    payload.identifier = fieldIdentifier;
+    payload.location = locality;
+    payload.area = Number(area);
+    payload.id = selectedRow.obj.id;
+    
+    // console.log(selectedRow.obj);
+    if (operation === 'UPDATE') {
+      updateField(payload)
+        .then((response) => {
+          // showAlert('Field successfully updated', 'info');
```
Comment (gautamguptabasant) [inline src/components/common/FieldForm.js]: Please cleanup the comments

File: src/components/common/FieldForm.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+          closeHandler();          
+        })
+        .catch((e) => {
+          console.log('Internal server error', e);
+          showAlert('Field updation failed: ' + e.message, 'error');
+        });
+    } else {
+      createField(payload)
```
Comment (gautamguptabasant) [inline src/components/common/FieldForm.js:92]: Let's remove this for now and add later if required/used.

File: src/components/common/FieldForm.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    whiteSpace: 'normal',
+  },
+  selectEmpty: {
+    marginTop: theme.spacing(2),
+  },
+}));
+
+const FieldForm = ({ operation, title, selectedRow, closeHandler }) => {
+  const classes = useStyles();
+  const [locality, setLocality] = useState(selectedRow.locationId);
```
Comment (gautamguptabasant) [inline src/components/common/FieldForm.js]: This state stores locationCode. Let's use this name.

File: src/components/common/FieldForm.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+          showAlert('Field successfully created', 'info');
+        })
+        .catch((e) => {
+          console.log('Internal server error', e);
+          showAlert('Field creation failed: ' + e.message, 'error');
+        });
+    }
+  };
+
+  const handleChange = ({ target }) => {
```
Comment (gautamguptabasant) [inline src/components/common/FieldForm.js:103]: This method could be merged with `handleTextChange`. Doing the same thing in both.

File: src/components/common/FieldForm.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+      />
+
+      <Button
+        variant='contained'
+        color='primary'
+        size='medium'
+        className={classes.submitButton}
+        onClick={handleSubmit}
+      >
+        Submit
```
Comment (gautamguptabasant) [inline src/components/common/FieldForm.js]: Let's take this also as a prop in the component. We can use `Save` for update and `Create` for creation.

File: src/components/common/SimpleModal.js [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+function getModalStyle() {
+  const top = 50 + rand();
+  const left = 50 + rand();
+
+  return {
+    top: `${top}%`,
+    left: `${left}%`,
+    transform: `translate(-${top}%, -${left}%)`,
+  };
+}
```
Comment (gautamguptabasant) [inline src/components/common/SimpleModal.js]: I think you need to remove this and set the margins to auto to get the modal in centre of the page always.

File: src/components/common/SimpleModal.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+}));
+
+export default function SimpleModal({
+  isOpen,
+  closeHandler,
+  modalBody
+}) {
+  const classes = useStyles();
+  // getModalStyle is not a pure function, we roll the style only on the first render
+  const [modalStyle] = React.useState(getModalStyle);
```
Comment (gautamguptabasant) [inline src/components/common/SimpleModal.js]: Please add `useState` in the import to be consistent with the way we are importing stuff.

File: src/dataclients/FieldsClient.js [hunk 14]
```diff
@@ -38,3 +38,19 @@ export const createField = async (payload) => {
     .then(handleErrors)
     .then((response) => response.json());
 };
+
+/**
+ * API to create a new field
```
Comment (gautamguptabasant) [inline src/dataclients/FieldsClient.js]: create -> update

**Review by gautamguptabasant — CHANGES_REQUESTED** [review]

File: src/Pages/FieldInfoPage.js (hunk 5 above)
Comment (aashnaPrakash) [reply src/Pages/FieldInfoPage.js]: Any specific reason why do we need the entire object, we are already mapping all the fields needed.

File: src/components/common/FieldForm.js [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    }
+  };
+
+  const handleSubmit = () => {
+    let payload = {};
+    // payload.name = fieldIdentifier;
+    payload.identifier = fieldIdentifier;
+    payload.location = locality;
+    payload.area = Number(area);
+    payload.id = selectedRow.obj.id;
```
Comment (aashnaPrakash) [inline src/components/common/FieldForm.js]: selectedRow already has the Id, we do not need selectedRow.obj.id I guess

File: src/Pages/FieldInfoPage.js (hunk 1 above)
Comment (mahimavuppuluri) [reply src/Pages/FieldInfoPage.js:16]: Modal is the name given to the import from material ui, which is used in constructing the SimpleModal, so we need to give this a different name

File: src/Pages/FieldInfoPage.js (hunk 3 above)
Comment (mahimavuppuluri) [reply src/Pages/FieldInfoPage.js:111]: its there

File: src/Pages/FieldInfoPage.js [hunk 16]
```diff
@@ -173,6 +184,8 @@ function FieldInfoPage() {
    * @param {object} selectedRow
```
Comment (aashnaPrakash) [inline src/Pages/FieldInfoPage.js:183]: can we remove the //todo comment

File: src/components/common/FieldForm.js (hunk 8 above)
Comment (mahimavuppuluri) [reply src/components/common/FieldForm.js:92]: will be needed soon, since add field will use the same form

File: src/Pages/FieldInfoPage.js (hunk 6 above)
Comment (mahimavuppuluri) [reply src/Pages/FieldInfoPage.js:266]: lets do this later

File: src/Pages/FieldInfoPage.js (hunk 16 above)
Comment (mahimavuppuluri) [reply src/Pages/FieldInfoPage.js:183]: thanks, updated

File: src/components/common/FieldForm.js (hunk 10 above)
Comment (mahimavuppuluri) [reply src/components/common/FieldForm.js:103]: thanks, removed this method and used handleChange

**Review by gautamguptabasant — APPROVED** [review]

**Pull request merged**